File: Retos/Reto #24 - CIFRADO CÉSAR [Fácil]/ruby/test0n3.rb
# frozen_string_literal: true

# class CaesarCypher
class CaesarCipher
  attr_accessor :input

  ALPHABET = { 'a' => 0, 'b' => 1, 'c' => 2, 'd' => 3, 'e' => 4, 'f' => 5, 'g' => 6, 'h' => 7,
               'i' => 8, 'j' => 9, 'k' => 10, 'l' => 11, 'm' => 12, 'n' => 13, 'ñ' => 14, 'o' => 15,
               'p' => 16, 'q' => 17, 'r' => 18, 's' => 19, 't' => 20, 'u' => 21, 'v' => 22,
               'w' => 23, 'x' => 24, 'y' => 25, 'z' => 26 }.freeze

  SWIFT_DIRECTIONS = { 'positive' => 1, 'negative' => -1 }.freeze

  def initialize(input = { text: nil, action: nil })
    @input = input
    @text = input[:text]
    @action = input[:action]
    @swift_direction = SWIFT_DIRECTIONS['positive']
    @swift_number = 3
  end

  def act
    return 'input missing' if @action.nil? || @input.nil?

    process_text
  end

  def process_text
    @text.chars.map do |letter|
      pos = ALPHABET[letter.downcase]
      if pos.nil?
        letter
      else
        new_value = ALPHABET.key(new_position(pos)).dup
        upcase?(letter) ? new_value.upcase! : new_value
      end
    end.join
  end

  private

  def upcase?(character)
    character == character.upcase
  end

  def new_position(position)
    offset = @swift_direction * @swift_number
    offset *= -1 if @action == 'decipher'

    (position + offset) % ALPHABET.size
  end
end

# puts CaesarCipher.new({ text: 'The quick brown fox jumps over the lazy dog.', action: 'cipher' }).act
# puts CaesarCipher.new({ text: 'Wkh txlfn eurzp ira mxosv ryhu wkh ñdcb grj.', action: 'decipher' }).act
# puts CaesarCipher.new.act

File: Retos/Reto #24 - CIFRADO CÉSAR [Fácil]/javascript/test0n3.js
function CaesarCipher(str = "", action = "") {
  const alphabet = {
    a: 0,
    b: 1,
    c: 2,
    d: 3,
    e: 4,
    f: 5,
    g: 6,
    h: 7,
    i: 8,
    j: 9,
    k: 10,
    l: 11,
    m: 12,
    n: 13,
    ñ: 14,
    o: 15,
    p: 16,
    q: 17,
    r: 18,
    s: 19,
    t: 20,
    u: 21,
    v: 22,
    w: 23,
    x: 24,
    y: 25,
    z: 26,
  };

  const swiftDirections = { positive: 1, negative: -1 };

  let text = str.split("");
  const swiftDirection = swiftDirections["positive"];
  const swiftNumber = 3;
  const alphabetLength = Object.keys(alphabet).length;

  function act() {
    if (str.length == 0 || action.length == 0) {
      return "input missing";
    }
    return process_text();
  }

  function process_text() {
    let result = [];
    text.forEach((letter) => {
      let pos = alphabet[letter.toLowerCase()];
      if (pos == undefined) {
        result.push(letter);
      } else {
        let newLetter = Object.keys(alphabet)[newPosition(pos)];
        if (isUpcase(letter)) {
          result.push(newLetter.toUpperCase());
        } else {
          result.push(newLetter);
        }
      }
    });
    return result.join("");
  }

  function isUpcase(letter) {
    return letter == letter.toUpperCase();
  }

  function newPosition(position) {
    let offset = swiftDirection * swiftNumber;
    if (action == "decipher") {
      offset = -offset;
    }
    let newPosition = (position + offset) % alphabetLength;
    if (newPosition < 0) {
      newPosition = alphabetLength + newPosition;
    }
    return newPosition;
  }

  Object.defineProperty(this, "act", {
    get: function () {
      return act();
    },
  });
}

const testsCollections = {
  tests: [
    { str: "The quick brown fox jumps over the lazy dog.", action: "cipher" },
    { str: "Wkh txlfn eurzp ira mxosv ryhu wkh ñdcb grj.", action: "decipher" },
    { str: "Aquí va más texto.", action: "cipher" },
    { str: "Dtxí yd oáv whawr.", action: "decipher" },
    { str: "", action: "" },
  ],
  expecteds: [
    "Wkh txlfn eurzp ira mxosv ryhu wkh ñdcb grj.",
    "The quick brown fox jumps over the lazy dog.",
    "Dtxí yd oáv whawr.",
    "Aquí va más texto.",
    "input missing",
  ],
};

testsCollections.tests.forEach((test, index) => {
  const result = new CaesarCipher(test.str, test.action);
  console.log(result.act);
  console.log(result.act == testsCollections.expecteds[index]);
  console.log("----------------------------------------------------");
});
